frontend/console/src/features/traces/traces.utils.ts

```diff
@@ -16,3 +17,30 @@ export const eventBarLeftOffsetPercentage = (event: Event, requestStartTime: Tim
 
   return (offsetInMillis / requestDurationMs) * 100
 }
+
+export const groupEventsByRequestKey = (events: Event[]): Record<string, Event[]> => {
+  const even = events.reduce((acc: Record<string, Event[]>, event: Event) => {
+    let requestKey: string | undefined
+
+    if (event.entry.case === 'call') {
+      requestKey = event.entry.value.requestKey
+    } else if (event.entry.case === 'ingress') {
+      requestKey = event.entry.value.requestKey
+    }
+
+    if (requestKey) {
+      if (!acc[requestKey]) {
+        acc[requestKey] = []
+      }
+
+      acc[requestKey].push(event)
+
+      // first event will be the "trigger" event
+      acc[requestKey].sort((a, b) => compareTimestamps(a.timeStamp, b.timeStamp))
+    }
+
+    return acc
+  }, {})
+
+  return even
+}
```

Comment:
Yeah, your version looks much better :)

There are a couple odd things to deal with with the generated `oneof` code from protobufs and from the `requestKey` being optional on events. But I think we can def use some of these patterns here.

Also added a switch because we'll have more types in the future.

```ts
export const groupEventsByRequestKey = (events: Event[]): Record<string, Event[]> => {
  return events.reduce((acc: Record<string, Event[]>, event: Event) => {
    let requestKey: string | undefined

    switch (event.entry.case) {
      case 'call':
      case 'ingress':
        requestKey = event.entry.value.requestKey
        break
      default:
        return acc
    }

    if (!requestKey) {
      return acc
    }

    acc[requestKey] = acc[requestKey] ? [...acc[requestKey], event] : [event]

    // Sort events by timestamp, ensuring the first event is the "trigger" event
    acc[requestKey].sort((a, b) => compareTimestamps(a.timeStamp, b.timeStamp))

    return acc
  }, {})
}
```

Reply:
Ah gotcha. The switch is awesome! With the `default: return acc`, can we remove the `if (!requestKey)` block below?

Reply:
I made it slightly better 😂 but with `requestKey` being optional even on those events, we still have to check it somewhere.

Reply:
ahhhhhh that's annoying, that makes sense